glean-core/src/scheduler.rs

```diff
@@ -0,0 +1,496 @@
+        );
+        schedule_internal(&glean, submitter, scheduler, fake_now);
+        assert_eq!(0, submitter_count.swap(0, Ordering::Relaxed));
+        assert_eq!(1, scheduler_count.swap(0, Ordering::Relaxed));
+    }
+
+    // If we haven't sent a ping today and we're after the scheduled time,
+    // ensure we send a ping and then schedule a "reschedule" ping for tomorrow.
+    // ("Case #2" in schedule_internal)
+    #[test]
+    fn test_case_2() {
+        let (glean, _t) = new_glean(None);
+
+        let fake_yesterday = FixedOffset::east(0)
+            .ymd(2021, 4, 29)
+            .and_hms(SCHEDULED_HOUR, 0, 1);
+        get_last_sent_time_metric().set(&glean, Some(fake_yesterday));
+        let fake_now = fake_yesterday + Duration::days(1);
+
+        let (submitter, submitter_count, scheduler, scheduler_count) = new_proxies(
+            |_, reason| assert_eq!(reason, Some("overdue")),
+            |_, when| assert_eq!(when, When::Reschedule),
+        );
+        schedule_internal(&glean, submitter, scheduler, fake_now);
+        assert_eq!(1, submitter_count.swap(0, Ordering::Relaxed));
+        assert_eq!(1, scheduler_count.swap(0, Ordering::Relaxed));
+    }
+
+    // If we haven't sent a ping today and we're before the scheduled time,
+    // ensure we don't send a ping but schedule a "today" ping for today.
+    // ("Case #3" in schedule_internal)
+    #[test]
+    fn test_case_3() {
+        let (glean, _t) = new_glean(None);
+
+        let fake_yesterday =
+            FixedOffset::east(0)
+                .ymd(2021, 4, 29)
+                .and_hms(SCHEDULED_HOUR - 1, 0, 1);
+        get_last_sent_time_metric().set(&glean, Some(fake_yesterday));
+        let fake_now = fake_yesterday + Duration::days(1);
+
+        let (submitter, submitter_count, scheduler, scheduler_count) = new_proxies(
+            |_, reason| panic!("Case #3 shouldn't submit a ping! reason: {:?}", reason),
+            |_, when| assert_eq!(when, When::Today),
+        );
+        schedule_internal(&glean, submitter, scheduler, fake_now);
+        assert_eq!(0, submitter_count.swap(0, Ordering::Relaxed));
+        assert_eq!(1, scheduler_count.swap(0, Ordering::Relaxed));
+    }
+
+    // `When` is responsible for date math. Let's make sure it's correct.
+    #[test]
+    fn test_when() {
+        let now = FixedOffset::east(0).ymd(2021, 4, 30).and_hms(15, 2, 10);
+        // `now` is after `SCHEDULED_HOUR` so should be zero:
+        assert_eq!(std::time::Duration::from_secs(0), When::Today.until(now));
+        // If we bring it back before `SCHEDULED_HOUR` it should give us the duration:
+        let earlier = now.date().and_hms(SCHEDULED_HOUR - 1, 0, 0);
+        assert_eq!(
+            std::time::Duration::from_secs(3600),
+            When::Today.until(earlier)
+        );
+
+        // `Tomorrow` and `Reschedule` should differ only in their `reason()`
+        assert_eq!(
+            std::time::Duration::from_secs(46670),
```

Comment:
I first was a bit confused by this "magic" duration of seconds. It's 12 hours, 57 minutes and 10 seconds. The exact time it takes from 15:02:10 to 00:04:00. Worth noting somewhere?